### ISSUE TITLE
A bug in MetricTracker with Lightning logging style

### ISSUE THREAD

**Issue by rustamzh** [Lightning-AI/metrics#1543]
## 🐛 Bug

Values logged by a standalone metric and the same metric in the MetricTracker differ, when logged using passing object style in  pytorch lightning logging.

### To Reproduce

```       
    def init()
        ...
        metrics = torchmetrics.MetricCollection([BinaryAccuracy(), BinaryAUROC(), BinaryAveragePrecision()])
        self.optimization_metric = MetricTracker(BinaryAveragePrecision())

        self.train_metrics = metrics.clone(prefix='train_')
        self.valid_metrics = metrics.clone(prefix='val_')
        self.test_metrics = metrics.clone(prefix='test_')
 
   def on_validation_epoch_start(self):
        self.optimization_metric.increment()

   def validation_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self(x)
        loss = self.criterion(y_hat, y.float())
        
        self.log('val_loss', loss)
        self.valid_metrics.update(y_hat, y)
        self.log_dict(self.valid_metrics, on_step=False, on_epoch=True)
        self.optimization_metric.update(y_hat, y)
        self.log('val_AP', self.optimization_metric.best_metric(), on_step=False, on_epoch=True)
```
Values in the csv logger:

| val_BinaryAveragePrecision | val_AP             |
|----------------------------|--------------------|
|         0.8018825650215149 | 0.7876660227775574 |

for the same epoch
### Expected behavior

Values in the logger should match each other in the first epoch. Or, there should be a warning that MetricTracker is incompatible with this logging style. When I changed the logging style, the metric values began to match.

### Environment

- TorchMetrics version - conda '0.11.1':
- Pytroch Lightning - conda '1.8.6'
- Python & PyTorch Version (e.g., 1.0): Python 3.10.9 & '1.13.1'
- Any other relevant information such as OS (e.g., Linux): Linux

**Comment by SkafteNicki** [comment]
Hi @rustamzh, thanks for raising this issue.
After investigation, I was able to reproduce your issue. However, this is simply a matter of how the logging interface should be called. 
I assume that when you call:
```python
self.log('val_AP', self.optimization_metric.best_metric(), on_step=False, on_epoch=True)
```
you actually just want to log the single best value at the end of the epoch? However, when setting `on_epoch=True` it will log a single value but it will be the accumulated value over all steps. See this page:
https://pytorch-lightning.readthedocs.io/en/latest/extensions/logging.html#automatic-logging
Instead what you most likely want to do is to implement the call as:
```python
def on_validation_epoch_end(self):
    self.log('val_AP', self.optimization_metric.best_metric())
```
such that the best metric is only logged once at the end of the epoch. If I do this, I get that the `val_BinaryAveragePrecision` value and `val_AP` value is the same after epoch 1.

**Comment by rustamzh** [comment]
Hi @SkafteNicki, I totally agree that it is an issue of an implementation. However, it is not clear from the reading of the docs. It is too easy to screw this up and no warning or mention of this in the documentation. 

I wish there would be either a warning raised when the MetricTracker is improperly used, or there would be a mention in the docs that not all classes that contain word `Metric` are actually metrics (inherit from `Metric`), and thus should be treat differently when used with the Lightning.

**Comment by SkafteNicki** [comment]
> I wish there would be either a warning raised when the MetricTracker is improperly used, or there would be a mention in the docs that not all classes that contain word `Metric` are actually metrics (inherit from `Metric`), and thus should be treat differently when used with the Lightning.

I do not think there is a simple way for us to add warning that will only be triggered in this specific case, but you are totally right that we should such information to the documentation. Will do that :]

**Comment by rustamzh** [comment]
Thanks!
